### PR TITLE
fix(fetch-http-handler): handle empty blob in browsers

### LINKED ISSUE #1198: Empty blob decoding via base64 is causing issues in Chrome
**Describe the bug**
Empty blob decoding via base64 is causing issues in Chrome

**SDK version number**
Gamma.1

**Is the issue in the browser/Node.js?**
Browser: Chrome Version 81.0.4044.138

**To Reproduce (observed behavior)**
Make any service call that uses the stream collector and returns empty results. S3 PUT reproduces it easily. See https://github.com/aws-amplify/amplify-js/issues/5876 and https://github.com/aws-amplify/amplify-js/issues/5873

**Additional context**
The issue seems to be how browsers implement `FileReader.readAsDataURL()`. On firefox, for an empty blob (case for a successful S3 PUT request) I'm seeing `FileReader.readAsDataURL()` returns `data:application/octet-stream;base64,` while on chrome it returns `data:`. This causes issues with parsing the actual result [here](https://github.com/aws/aws-sdk-js-v3/blob/master/packages/fetch-http-handler/src/stream-collector.ts#L42-L48). On chrome when it doesn't find a `,` it leaves the input as is which later cannot be decoded since `:` is not a valid base64 character.

Both IE and edge seems to return `null` for empty blobs so it shouldn't create any issues.

Tested with the following code in the 4 browsers
```
var testBlob = new Blob([]) // and again with non empty Blob(["a"])
var fileReader = new FileReader();
fileReader.onloadend = () => { console.log(fileReader.result) }
fileReader.readAsDataURL(testBlob);
```
Resulted in following observations
Blob | Firefox(72.0.2) | Chrome (81) | Edge(44) | IE (11)
------------ | ------------ | -------------  | ------------ | -------------
Empty | data:application/octet-stream;base64, | data: | null | null
Non empty | data:application/octet-stream;base64,YQ== | data:application/octet-stream;base64,YQ== | data:;base64,YQ== | data:;base64,YQ==

### PR DESCRIPTION
**Pull request by AllanFly120**
Fixes #1198 

This PR fixes the issue with returning none base64 encoded data for empty blob in browser. 

The original `readToBase64()` in `streamCollector` from `fetch-http-handler` was migrated from `stream-collector-native` package which was only intended to be used in ReactNative environment. It handles empty blobs fine in ReactNative, but misses unit test for the case. When the `stream-collector-native` is merged into `fetch-http-handler`, it can also be used in browser runtime, which can have different behavior from different browser vendors. This results untested behavior in browser runtime. 

/cc @Amplifiyer 

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by AllanFly120** [comment]
Tested with with Amplify Storage component(code snippet below) locally on Firefox and Chrome:
```javascript
  const { Storage, Amplify } = require("aws-amplify");
  Amplify.configure({
    Storage: {
      AWSS3: {
        bucket: "bucket",
        region,
        identityPoolId: IDENTITY_POOL_ID,
      },
    },
  });
  await Storage.put("abcd");
```

File: packages/fetch-http-handler/src/stream-collector.ts [hunk 1]
```diff
@@ -45,7 +45,10 @@ function readToBase64(blob: Blob): Promise<string> {
         return reject(new Error("Reader aborted too early"));
       }
       const result = reader.result as string;
-      const dataOffset = result.indexOf(",") + 1;
```
Comment (trivikr) [inline packages/fetch-http-handler/src/stream-collector.ts:48]: need to add `null` check as null.indexOf will throw
`Cannot read property 'indexOf' of null`

File: packages/fetch-http-handler/src/stream-collector.spec.ts [hunk 2]
```diff
@@ -29,4 +29,16 @@ describe("streamCollector", () => {
...
   });
+
+  it("returns a Uint8Array when blob is empty", done => {
+    const expected = new Uint8Array(0);
+
+    const dataPromise = new Response(expected.buffer)
+      .blob()
+      .then(blob => streamCollector(blob));
+    dataPromise.then((data: any) => {
+      expect(data).toEqual(Uint8Array.from([]));
```
Comment (trivikr) [inline packages/fetch-http-handler/src/stream-collector.spec.ts:40]:
```suggestion
      expect(data).toEqual(expected);
```

File: packages/fetch-http-handler/src/stream-collector.ts [hunk 3]
```diff
@@ -44,8 +44,11 @@ function readToBase64(blob: Blob): Promise<string> {
       if (reader.readyState !== 2) {
         return reject(new Error("Reader aborted too early"));
       }
-      const result = reader.result as string;
-      const dataOffset = result.indexOf(",") + 1;
+      const result = (reader.result as string) || "";
```
Comment (trivikr) [inline packages/fetch-http-handler/src/stream-collector.ts:47]: Although TypeScript shows result can be `string | ArrayBuffer | null`, we're checking only for string as null as the operation name called is `readAsDataURL` on line 57  

<details>
<summary>Screenshot</summary>

Screen Shot 2020-05-26 at 11 02 07 AM

</details>

Comment (trivikr) [inline packages/fetch-http-handler/src/stream-collector.ts:47]:
```suggestion
      const result = (reader.result ?? "") as string;
```

TypeScript playgrounds:
* [Calling as string on result](https://www.typescriptlang.org/play?ts=3.8.3#code/C4TwDgpgBAThDOBXANsAKuaBeK9gwEsA7AcygB8oiVkBuAKHoGMB7IvKVogNwhnRYBlfMTI4AFMBYAhCAGE2vfhAAmALlgIU6TAEoNeQqShYAfFCmyFPPsFVQAhvFwjj5SgCIPDZottohVxJxD0MPXQYuJQFhI2DqZGQI+iA)
* [Calling as string on result ?? ""](https://www.typescriptlang.org/play?ts=3.8.3#code/C4TwDgpgBAThDOBXANsAKuaBeK9gwEsA7AcygB8oiVkBuAKHoGMB7IvKVogNwhnRYBlfMTI4AFMBYAhCAGE2vfhAAmALlgIU6TAEoNeQqShYAfFCmyFPPsFVQA-A6gAiFw2aLbaISNLiXQxddBi4lAWEjEnFqZGQQ+iA)

File: packages/fetch-http-handler/src/stream-collector.jest.spec.ts [hunk 4]
```diff
@@ -0,0 +1,36 @@
+import { streamCollector } from "./stream-collector";
```
Comment (trivikr) [inline packages/fetch-http-handler/src/stream-collector.jest.spec.ts:1]: retain name as `stream-collector.spec.ts`

File: packages/fetch-http-handler/src/stream-collector.spec.ts [hunk 5]
```diff
@@ -29,4 +29,16 @@ describe("streamCollector", () => {
       done();
     });
   });
+
+  it("returns a Uint8Array when blob is empty", done => {
```
Comment (trivikr) [inline packages/fetch-http-handler/src/stream-collector.spec.ts:33]: rename to `stream-collector.browser.spec.ts `

File: packages/fetch-http-handler/src/stream-collector.ts (hunk 1 above)
Comment (trivikr) [reply packages/fetch-http-handler/src/stream-collector.ts:48]: fixed in [commit](https://github.com/aws/aws-sdk-js-v3/pull/1199/commits/1e5fa454a42f0219eca536d2356f7191804cabb0)

File: packages/fetch-http-handler/src/stream-collector.spec.ts (hunk 2 above)
Comment (trivikr) [reply packages/fetch-http-handler/src/stream-collector.spec.ts]: fixed in [commit](https://github.com/aws/aws-sdk-js-v3/pull/1199/commits/1e5fa454a42f0219eca536d2356f7191804cabb0)

**Pull request merged**